runtime/lua/vim/diagnostic.lua

```diff
@@ -60,7 +60,24 @@ local bufnr_and_namespace_cacher_mt = {
 }
 
 local diagnostic_cleanup = setmetatable({}, bufnr_and_namespace_cacher_mt)
-local diagnostic_cache = setmetatable({}, bufnr_and_namespace_cacher_mt)
+local diagnostic_cache = setmetatable({}, {
+  __index = function(t, bufnr)
```

Comment:
So this is weird; you'd think `__newindex` would be the way to go here, but it never seems to be called, either here or with `bufnr_and_namespace_cacher_mt`. Maybe `__index` doing the `rawset(t, bufnr, {})` part has something to do with it? Not sure, but it seems weird.

Also this `__index` is basically exactly the same as the one in `bufnr_and_namespace_cacher_mt`, any thoughts on reducing the repetition here, or is it not that big of a deal?